README.md
# NestJS NuxtJs Boilerplate

[![Build Status](https://drone.zebbra.ch/api/badges/zebbra-repos/nestjs-nuxtjs-boilerplate/status.svg)](https://drone.zebbra.ch/zebbra-repos/nestjs-nuxtjs-boilerplate)
[![Automated Release Notes by gren](https://img.shields.io/badge/%F0%9F%A4%96-release%20notes-00B2EE.svg)](https://github-tools.github.io/github-release-notes/)

> NestJS and NuxtJS starter boilerplate

## Requirements

- Node.js 14.4.0
- PostgreSQL

## Development

For development the following additional dependencies are required:

- Homebrew (for Mac OS X)
- overmind
- yarn

### Setup

Make sure you have the latest version of `nodenv` (Node.js version manager) by following instruction:

```bash
brew upgrade nodenv
```

Create `development` and `test` databases:

```bash
CREATE DATABASE nest-nuxt-boilerplate-development
CREATE DATABASE nest-nuxt-boilerplate-test
```

Run migrations:

```bash
yarn db:migrate
NODE_ENV=test yarn db:migrate
```

### Project structure

| Folder       | Description                                                                                                                                       |
| ------------ | ------------------------------------------------------------------------------------------------------------------------------------------------- |
| `/`          | Contains project configuration files as well as typescript and linting specifications which are applied to both (NestJS and NuxtJS) applications. |
| `app/client` | Contains the NuxtJS application directory as described [here](https://nuxtjs.org/guide/directory-structure).                                      |
| `app/server` | Contains the NestJS application directory as described [here](https://docs.nestjs.com/first-steps).                                               |

### What's included

#### NuxtJS Frontend

- [@nuxtjs/pwa](https://pwa.nuxtjs.org) Supercharge Nuxt with a heavily tested, updated and stable PWA solution
- [@nuxtjs/stylelint-module](https://github.com/nuxt-community/stylelint-module) Stylelint module for NuxtJS
- [@nuxtjs/vuetify](https://github.com/nuxt-community/vuetify-module) Vuetify Module for NuxtJS
- [@nuxt/typescript-build](https://typescript.nuxtjs.org/guide/setup.html#installation) NuxtJS TypeScript Build Support
- [@nuxt/typescript-runtime](https://typescript.nuxtjs.org/guide/runtime.html#installation) NuxtJS TypeScript Runtime Support
- [@nuxtjs/composition-api](https://composition-api.now.sh/) Use Vue 3 Composition API with Nuxt-specific features
- [vuex-module-decorators](https://github.com/championswimmer/vuex-module-decorators) TypeScript/ES7 Decorators to create Vuex modules declaratively
- [@sentry/browser](https://www.npmjs.com/package/@sentry/browser) Sentry integration for browser
- [@nuxtjs/apollo](https://github.com/nuxt-community/apollo-module) Nuxt.js module to use vue-apollo
- [@vue/apollo-composable](https://v4.apollo.vuejs.org/guide/) This library integrates apollo in your Vue components with declarative queries

##### Also interesting for NuxtJS

- [@nuxtjs/style-resources](https://github.com/nuxt-community/style-resources-module#readme) Handling scss, sass, and stylus files
- [nuxt-webfontloader](https://github.com/Developmint/nuxt-webfontloader#readme) Efficient web font loading
- [@aceforth/nuxt-optimized-images](https://aceforth.com/docs/nuxt-optimized-images/) Automatically optimizes images used in NuxtJS projects (JPEG, PNG, SVG, WebP and GIF)

#### NestJS Backend

- [nestjs-pino](https://github.com/iamolegga/nestjs-pino) Platform agnostic logger for NestJS based on Pino with request context in every log
- [@nestjs/config](https://github.com/nestjs/config) Configuration module based on the dotenv package
- [@nestjs/graphq](https://github.com/nestjs/graphql) GraphQL (TypeScript) module for Nest framework
- [@nestjs/passport](https://github.com/nestjs/passport) Passport utilities module for Nest
- [@nestjs/typeorm](https://github.com/nestjs/typeorm) Typeorm module for Nest
- [@nestjs/terminus](https://www.npmjs.com/package/@nestjs/terminus) Integrated healthchecks for Nest
- [@ntegral/nestjs-sentry](https://www.npmjs.com/package/@ntegral/nestjs-sentry) Sentry module for NestJS

## Production Setup for k8s cluster

> This project is managed with helm v3

- Replace `nest-nuxt-boilerplate` with your `awesome-project-name` in `/deploy`
- Replace `your-namespace` with your own k8s namespace

1. Create shared namespace

   ```bash
   kubectl create ns your-namespace
   ```

1. Copy `deploy/secrets.example.yaml` to

   - deploy/secrets.yaml

1. Deploy secrets

   ```bash
   kubectl apply -n your-namespace -f deploy/secrets.yaml
   ```

1. Deploy Helm chart

   ```bash
   cd deploy && helmfile apply
   ```

## Deployment

Push to master branch will trigger new deployment on k8s.

## Build Setup

```bash
# install dependencies
$ yarn install

# serve with hot reload at localhost:3000 (backend) and localhost:5000 (frontend)
$ overmind s

# build for production and launch combinded server (NestJS server serves NuxtJS frontend)
$ yarn build
$ yarn start:prod
```

For detailed explanation on how things work, check out [NestJS docs](https://docs.nestjs.com/) or [NuxtJS docs](https://nuxtjs.org).

## DB Migrations

k8s deployment includes db-migrate-job which will automatically run migrations on each deplyoment. Locally you have to use the following commands to interact with typeorm cli.

```bash
yarn db:migration:generate <migration_name>
yarn db:migrate
yarn db:rollback
```

> To run migrations against test database use `NODE_ENV=test yarn db:migrate`

## Graphql Schema and Operations

We chose to use the Nest.js `code first` approach (see [here](https://docs.nestjs.com/graphql/quick-start)) which means that the `app/server/schemal.gql` file is automatically generated based on the types and resolvers we define in our Nest.js application. Further, we make use of [GraphQL Code Generator](https://graphql-code-generator.com/) to generate types and queries on the fly for the client based on the server's `schema.gql` file and the operations defined in one of the following files:

- `app/client/apollo/mutations.graphql`
- `app/client/apollo/queries.graphql`
- `app/client/apollo/subscriptions.graphql`

The results are written to `app/client/apollo/generated-operations.ts` and can be used as-is (see [here](https://graphql-code-generator.com/docs/plugins/typescript-vue-apollo) for more information).

If you run the application with `overmind s` then `generated-operations.ts` will be re-generated if there are changes to one of the files listed above. Otherwise you have to run `yarn generate:types` manually to re-generate the output file.


CHANGELOG.md
# Changelog

## v1.0.0 (11/09/2020)

### closed

- [**closed**] Switch to SSR instead of full static [#14](https://github.com/zebbra-repos/nestjs-nuxtjs-boilerplate/pull/14)
- [**closed**] Implement csrf protection and some refactoring [#13](https://github.com/zebbra-repos/nestjs-nuxtjs-boilerplate/pull/13)
- [**closed**] WIP: Feature/apollo client [#11](https://github.com/zebbra-repos/nestjs-nuxtjs-boilerplate/pull/11)
- [**closed**] Use nuxtjs static build and nestjs serve-static module [#3](https://github.com/zebbra-repos/nestjs-nuxtjs-boilerplate/pull/3)

### dependencies

- [**dependencies**] Bump lodash from 4.17.15 to 4.17.19 [#4](https://github.com/zebbra-repos/nestjs-nuxtjs-boilerplate/pull/4)

### enhancement

- [**enhancement**] WIP: Full Example [#5](https://github.com/zebbra-repos/nestjs-nuxtjs-boilerplate/pull/5)


nuxt.config.ts
import { NuxtConfig } from "@nuxt/types";

const config: NuxtConfig = {
  // Target: https://go.nuxtjs.dev/config-target
  // target: "server",
  // ssr: false,

  // Global page headers: https://go.nuxtjs.dev/config-head
  head: {
    title: process.env.npm_package_name || "",
    meta: [
      { charset: "utf-8" },
      { name: "viewport", content: "width=device-width, initial-scale=1" },
      {
        hid: "description",
        name: "description",
        content: process.env.npm_package_description || "",
      },
    ],
    link: [{ rel: "icon", type: "image/x-icon", href: "/favicon.ico" }],
  },

  // Customize the progress-bar color
  loading: { color: "#fff" },

  // Global CSS: https://go.nuxtjs.dev/config-css
  css: [],

  // Plugins to run before rendering page: https://go.nuxtjs.dev/config-plugins
  plugins: [
    {
      src: "~/plugins/apollo.plugin",
      mode: "all",
    },
    {
      src: "~/plugins/persistence.plugin",
      mode: "client",
    },
    {
      src: "~/plugins/settings.plugin",
      mode: "client",
    },
    {
      src: "~/plugins/sentry.plugin",
      mode: "all",
    },
    {
      src: "~/plugins/validator.plugin",
      mode: "all",
    },
    {
      src: "~/plugins/auth.plugin",
      mode: "client",
    },
  ],

  // Auto import components: https://go.nuxtjs.dev/config-components
  components: true,

  // Modules for dev and build (recommended): https://go.nuxtjs.dev/config-modules
  buildModules: [
    // https://typescript.nuxtjs.org/guide/setup.html#installation
    "@nuxt/typescript-build",
    // https://github.com/nuxt-community/stylelint-module
    "@nuxtjs/stylelint-module",
    // https://github.com/nuxt-community/vuetify-module
    "@nuxtjs/vuetify",
    // https://composition-api.nuxtjs.org/
    "@nuxtjs/composition-api/module",
  ],

  // Modules: https://go.nuxtjs.dev/config-modules
  modules: [
    // https://pwa.nuxtjs.org/
    ["@nuxtjs/pwa", { meta: false, icon: false, manifest: false }],
    // https://github.com/nuxt-community/apollo-module
    "@nuxtjs/apollo",
    // https://i18n.nuxtjs.org/
    "nuxt-i18n",
  ],

  // https://github.com/nuxt-community/apollo-module
  apollo: {
    clientConfigs: {
      default: "~/apollo/config/default",
    },
  },

  // https://github.com/nuxt-community/vuetify-module
  vuetify: {
    customVariables: ["~/assets/stylesheets/variables.scss"],
    treeShake: true,
    optionsPath: "~/vuetify.options.ts",
  },

  // https://i18n.nuxtjs.org/
  i18n: {
    locales: [
      {
        code: "en",
        file: "en.js",
      },
      {
        code: "de",
        file: "de.js",
      },
    ],
    defaultLocale: "en",
    strategy: "no_prefix",
    lazy: true,
    langDir: "locales/",
    detectBrowserLanguage: {
      fallbackLocale: "en",
      onlyOnRoot: true,
    },
    vueI18n: {
      fallbackLocale: "en",
    },
  },

  // PWA module configuration: https://go.nuxtjs.dev/pwa
  pwa: {
    manifest: {
      lang: "en",
    },
  },

  // https://nuxtjs.org/api
  srcDir: "app/client",
  buildDir: "dist/app/client",
  dir: {
    // Static serve dir is resolved to path.resolve(srcDir, dir.static)
    // See https://github.com/nuxt/nuxt.js/blob/dev/packages/server/src/server.js
    static:
      process.env.NODE_ENV === "production"
        ? "../../dist/app/client/dist/static"
        : "static",
  },
  server: {
    port: process.env.PORT || 5000,
  },

  // Runtime config: https://nuxtjs.org/docs/2.x/configuration-glossary/configuration-runtime-config
  publicRuntimeConfig: {
    apollo: {
      httpGraphQLEndpoint:
        process.env.HTTP_GRAPHQL_ENDPOINT || "http://localhost:3000/graphql",
      wsGraphQLEndpoint:
        process.env.WS_GRAPHQL_ENDPOINT || "ws://localhost:3000/graphql",
    },
    sentry: {
      dsn: process.env.SENTRY_DSN,
      environment: process.env.SENTRY_ENVIRONMENT,
      logErrors: process.env.NODE_ENV !== "production",
    },
  },

  // Build Configuration: https://go.nuxtjs.dev/config-build
  build: {
    extend(config) {
      config.node = {
        fs: "empty",
      };
    },
  },

  // Do not send anonymous telemetry data: https://nuxtjs.org/docs/2.x/configuration-glossary/configuration-telemetry
  telemetry: false,
};

export default config;
